<div class="alert alert-block alert-info"><b></b>
<h1><center> <font color='black'> Homework 01  </font></center></h1>
<h2><center> <font color='black'> Introduction and first look at the data </font></center></h2>    
<h2><center> <font color='black'> BDA - University of Tartu - Spring 2020</font></center></h3>
</div>


# 1. Python and Pandas (4.5 points)

* In this course we are going to use Python, and mainly Pandas data structures and data analysis tool. please make sure that you have went through the practise session carefully. After that please answer the following questions below about the abalone dataset.

First, let's import pandas package as pd and read data from the CSV file into a pandas DataFrame.

In [1]:
import pandas as pd
data = pd.read_csv("abalone.csv")
data

Gender  Length  Diameter  Height  Weight  Rings
0        F   0.505     0.385   0.135  0.6185     12
1        F   0.650     0.475   0.165  1.3875      9
2        I   0.520     0.380   0.135  0.5395      8
3        F   0.550     0.425   0.140  0.9520      7
4        M   0.640     0.525   0.180  1.3135     10
..     ...     ...       ...     ...     ...    ...
995      I   0.550     0.415   0.135  0.8095     12
996      I   0.455     0.435   0.110  0.4265      8
997      M   0.660     0.500   0.165  1.3195      9
998      I   0.525     0.400   0.130  0.6455      8
999      M   0.515     0.395   0.135  1.0070      8

[1000 rows x 6 columns]

`Question 0` is an example of how you should answer the following questions. First, in the `#TODO` cell you have to write the code representing your solution. And in the answer cell write the answer with your own words. It does not have to be a full sentence.

 **1.0. What is overall memory consumption of the data?** (EXAMPLE)

In [2]:
# TODO - here you have to write the code how did you find the answer.
data.memory_usage(deep=True).sum()

102128

**<font color='red'>Answer:</font>** The total memory usage is 102080 bytes (102,08 kb).

 **1.1. What are the column names of the dataset?** **(0.25 point)**

In [3]:
data.columns

Index(['Gender', 'Length', 'Diameter', 'Height', 'Weight', 'Rings'], dtype='object')

**<font color='red'>Answer:</font>** 'Gender', 'Length', 'Diameter', 'Height', 'Weight', 'Rings'

**1.2. How many observations (i.e. rows) are in this data frame?** **(0.25 point)**

In [4]:
data.index

RangeIndex(start=0, stop=1000, step=1)

**<font color='red'>Answer:</font>** 1000

**1.3. Print the first 5 lines from the dataset. What are the values of feature "Rings" of the printed observations?** **(0.25 point)**

In [5]:
data.head()

Gender  Length  Diameter  Height  Weight  Rings
0      F   0.505     0.385   0.135  0.6185     12
1      F   0.650     0.475   0.165  1.3875      9
2      I   0.520     0.380   0.135  0.5395      8
3      F   0.550     0.425   0.140  0.9520      7
4      M   0.640     0.525   0.180  1.3135     10

**<font color='red'>Answer:</font>** 12, 9, 8, 7, 10

**1.4. Extract the last 3 rows of the data frame. What is the "Weight" of these abalones?** **(0.25 point)**

In [6]:
data.tail(3).Weight

997    1.3195
998    0.6455
999    1.0070
Name: Weight, dtype: float64

**<font color='red'>Answer:</font>** 1.3195, 0.6455, 1.0070

**1.5. What is the value of diameter in the row 577?** **(1 point)**

In [7]:
data.loc[576]['Diameter']

0.51

**<font color='red'>Answer:</font>** 0.51

**1.6. What is the mean of the height column?** **(0.5 point)** <br>
Hint 1: Use pandas describe() method

In [8]:
data['Height'].describe()

count    1000.000000
mean        0.141100
std         0.057064
min         0.000000
25%         0.115000
50%         0.145000
75%         0.165000
max         1.130000
Name: Height, dtype: float64

**<font color='red'>Answer:</font>** 0.141100

**1.7. Extract the subset of rows of the data frame where gender is M and weight values are below 0.75. What is the mean of diameter in this subset?** **(1 point)**

In [13]:
data[(data['Gender']=="M")&(data['Weight']<0.75)].describe()

Length    Diameter      Height      Weight       Rings
count  119.000000  119.000000  119.000000  119.000000  119.000000
mean     0.444118    0.342647    0.115504    0.467366    8.865546
std      0.075819    0.063027    0.022193    0.196004    1.986925
min      0.235000    0.160000    0.055000    0.051500    4.000000
25%      0.385000    0.295000    0.100000    0.291500    7.500000
50%      0.465000    0.360000    0.115000    0.495500    9.000000
75%      0.500000    0.385000    0.130000    0.643250   10.000000
max      0.575000    0.455000    0.190000    0.749500   14.000000

In [2]:
data.index='length', level=1

SyntaxError: can't assign to literal (<ipython-input-2-1e36e5c5507b>, line 1)

**<font color='red'>Answer:</font>** 0.342647

**1.8. What is the minimum of length when rings is equal to 18?** **(1 point)**

In [ ]:
data[data['Rings']==18].describe()

**<font color='red'>Answer:</font>** 0.465

# 2.  Data understanding (2.5 point)

Collecting, describing, exploring and verifying data quality are <b>very important</b> steps in any data science task. In this exercise the main goal is to understand the attributes in the 'adult' dataset.<br>

**2.0. Load data from the  `adult.csv` file.** <br>
Hint 1: The extension of the filename may not always reflect the structure of data precisely. Are the values comma-separated (csv) or not? 

In [3]:
import numpy as np
import pandas as pd

# TODO
data = pd.read_table('adult.csv',index_col='X')
data

age          workclass    education          occupation  capital.gain  \
X                                                                             
0      39          State-gov    Bachelors        Adm-clerical          2174   
1      50   Self-emp-not-inc    Bachelors     Exec-managerial             0   
2      38             privat      HS-grad   Handlers-cleaners             0   
3      53             privat         11th   Handlers-cleaners             0   
4      28             privat    Bachelors      Prof-specialty             0   
...    ..                ...          ...                 ...           ...   
32556  27            Private   Assoc-acdm        Tech-support             0   
32557  40            Private      HS-grad   Machine-op-inspct             0   
32558  58            Private      HS-grad        Adm-clerical             0   
32559  22            Private      HS-grad        Adm-clerical             0   
32560  52       Self-emp-inc      HS-grad     Exec-managerial         15024   

       capital.loss  native.country  salaries jobsatisfaction  male  female  
X                                                                            
0                 0   United-States   43136.0               0   1.0     NaN  
1                 0    UnitedStates   46209.0               7   1.0     NaN  
2                 0   United-States   28937.0               6   1.0     NaN  
3                 0   United-States   33658.0               6   1.0     NaN  
4                 0            Cuba   34372.0              12   NaN     1.0  
...             ...             ...       ...             ...   ...     ...  
32556             0   United-States   21393.0              14   NaN     1.0  
32557             0   United-States   75199.0              10   1.0     NaN  
32558             0   United-States   25603.0               1   NaN     1.0  
32559             0   United-States   42138.0               3   1.0     NaN  
32560             0   United-States   71709.0               9   NaN     1.0  

[32561 rows x 11 columns]

**2.1. Name all the columns in this dataset and describe the meaning of each column with one sentence.** <br>

In [ ]:
data.columns

**<font color='red'>Answer:</font>** 
-age: age of adult person
-workclass: type of work including: private, self-employee and state-gov
-education: latest education
-occupation: name of job
-capital gain: a rise in the value of a capital asset (investment or real estate) that gives it a higher worth than the purchase price [reference](https://www.investopedia.com/terms/c/capitalgain.asp)
-capital loss: the loss incurred when a capital asset, such as an investment or real estate, decreases in value [reference](https://www.investopedia.com/terms/c/capitalloss.asp)
-nativecountry: country which was born
-salaries: annual salary in USD
-jobsatisfication:Contentment (or lack of it) arising out of interplay of employee's positive and negative feelings toward his or her work [reference](http://www.businessdictionary.com/definition/job-satisfaction.html)  
-male: describe the gender is male or no
-female: describe the gender is female or no

####  2.1.1. In the following subtasks, you will have to visualise the distribution of the values for different features using appropriate diagrams. Visualisation is one of the easiest ways to get a good overview of the data and to notice any oddities, outliers etc.

In [ ]:
# Matplotlib is a Python 2D plotting library which produces publication quality figures in a variety of hardcopy formats 
# and interactive environments across platforms. Matplotlib can be used in Python scripts, the Python and IPython shells, 
# the Jupyter notebook, web application servers, and four graphical user interface toolkits.
import matplotlib.pyplot as plt

# allows to output plots in the notebook
%matplotlib inline 

# Set the default style
plt.style.use("ggplot") 

As an example, here is the age frequency histogram, where age is an attribute of type **numeric**:

In [ ]:
# Columns may contain non-numeric values, errors or missing values. Therefore, non-numeric values must be dealt with.
if data is not ...:
    pd.to_numeric(data['age'], errors='coerce').hist(bins=30) # ‘coerce’ -> invalid parsing will be set as NaN
else:
    print('Please define `data` in earlier subtasks')

**2.2. Attribute `Workclass`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** Nominal

In [ ]:
fig = plt.figure(figsize=[16, 8])

ax = plt.subplot(111)
ax.hist(data['workclass'],
        width = 0.5,
        weights = np.ones(len(data['workclass']))/ len(data['workclass'])
        )
ax.set_title('Workclass - Distribution in Relative Frequency')
plt.xticks(rotation = 90)
plt.show()

**2.3. Attribute `Education`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** ordinal.

In [ ]:
fig = plt.figure(figsize=[16, 8])

ax = fig.add_subplot(111)
ax.hist(data['education'],
        width = 0.5,
        weights = np.ones(len(data['education']))/ len(data['education'])
        )
ax.set_title('Relative frequency histogram')
plt.xticks(rotation = 90)
plt.show()

**2.4. Attribute `Occupation`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** nominal.

In [ ]:
fig = plt.figure(figsize=[16, 8])

ax = fig.add_subplot(111)
ax.hist(data['occupation'],
        width = 0.5,
        weights = np.ones(len(data['occupation']))/ len(data['occupation'])
        )
ax.set_title('Relative frequency histogram')
plt.xticks(rotation = 90)
plt.show()


**2.5. Attribute `Native Country`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** nominal

In [ ]:
fig = plt.figure(figsize=[16,8])

ax = fig.add_subplot(111)
ax.hist(data['native.country'],
        width = 1,
        weights = np.ones(len(data['native.country']))/ len(data['native.country'])
        )
ax.set_title('Relative frequency histogram')
plt.xticks(rotation = 90)
plt.show()

data


**2.6. Attribute `Salaries`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the density scale.**

**<font color='red'>Attribute type :</font>** numeric

In [ ]:
import seaborn as sns
sns.distplot(data['salaries'], hist=True, kde=True,
             bins=20, color = 'darkblue',
             hist_kws={'edgecolor':'black'},
             kde_kws={'linewidth': 2})



**2.7. Attributes `Male` and `Female`. First, specify the types of these attributes (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the relative frequency of positive values in attribute `Male` and positive values in attribute `Female` (relative frequency of males and females).**

**<font color='red'>Attribute type :</font>** nominal

In [ ]:
data['gender'] = data['male'].apply(lambda x: 'Male' if x == 1.0 else 'Female')

# TODO
fig = plt.figure(figsize=[6, 8])

ax4 = plt.subplot(111)
ax4.hist(data['gender'], 
         width = 0.5,
         weights = np.ones(len(data['gender'])) / len(data['gender']) 
        )
ax4.set_title('Gender - Distribution in Relative Frequency')
plt.xticks(rotation=90)
plt.show()

**2.8. Count the number of missing values in each attribute in the dataset and also report the percentage of missing values out of all rows (the original file uses '?' for missing values)**

In [ ]:
#TODO
total_rows = data.shape[0]
total_missing = 0
if data is not data.empty:
    for column in data.columns:
        total_missing = len(data[data[column] == '?'])
        if total_missing > 0:
            print (f"Total missing in column {column} = {total_missing}")
            print (f"Percentage of missing values in attribute {column} = {round((total_missing/total_rows)*100,2)}%")
else:
    print('Please define `data` in earlier subtasks')
    

# 3.  Data preparation (1.5 points)

Preparing the data for analysis is the most crucial phase in data analysis. In this exercise, the main goal is to clean up the dataset 'adult'. The dataset has been made unclean on purpose so that you can practice cleaning it up. Most of the introduced uncleanliness is of the same kinds as described in slides of Lecture 02. Additionally, some values have been introduced which are non-sensical due to the meaning of the attribute.<br>

First, let's make a deep copy of the original DataFrame. Later we will compare it to the cleaned/modified DataFrame.

In [ ]:
if data is not data.empty:
    original_data = data.copy(deep=True) # Make a deep copy, including a copy of the data and the indices

**3.1. Sometimes values include leading and/or trailing spaces. Remove leading and trailing whitespace from strings. Count how many rows were changed in the dataset.** <br>

In [ ]:
for column in data.columns:    
    if data[column].dtype == 'O':
        data[column] = data[column].str.strip()
    else:
        print(f"column {column} is not string")
        
data.head()

Use code below to count how many rows were affected.

In [ ]:
# Count differences
if data is not data.empty:
    data_all = pd.concat([original_data, data]).drop_duplicates()
    diff = data_all.shape[0] - data.shape[0]
    print ('Difference: ' + str(diff))

**<font color='red'>Answer:</font>**0

**3.2. Replace all non-sensical values in `data` with np.nan from numpy package to denote missing values (Original file uses '?' for missing values). Fix the typos in the dataset. Justify all the changes that you do, explaining why you are sure that this is a correct change and what you think the cause for such error could be.**  <br>

In [ ]:
data_clean = data.replace(['?','privat','UnitedStates', 'Unitedstates', 'Hong'], [np.nan,'Private','United-States', 'United-States', 'Hong-Kong'])
data_clean.head(10)


**<font color='red'>Answer:</font>** Beside replace `?` with `NaN` value, we also replace this string `['privat','UnitedStates', 'Unitedstates', 'Hong']` into `['Private','United-States', 'United-States', 'Hong-Kong']` to fix the typo and reduce the redudancy of the same value (for example: United-States and UnitedStates is the same value in `native.country`.

**3.3. Count the rows which were changed in the subtask b** <br>

In [ ]:
# TODO
if data is not data.empty:
    data_all = pd.concat([original_data, data_clean]).drop_duplicates()
    diff = data_all.shape[0] - data.shape[0]
    print ('Difference: ' + str(diff))

**<font color='red'>Answer:</font>** 4186

**3.4. Report values that look suspicious whereas you are not sure whether they are definitely wrong** <br> 

**<font color='red'>Answer:</font>** 
- value `Hong` in column `native.country`, we're not sure is it truly the name of the country `Hong Kong` or another meaning
- value in `capital.gain` and `capital.loss` which are containing too many `0` value
- value `Assoc-acdm` in `education` column

# 4.  Gathering interesting facts about the adult dataset (1.5 points)

Useful information about the data can be acquired through visualization even before modeling.

**4.1. Convert the attribute `Education` into type `category` and introduce the order for this attribute based on the educational level (e.g. Masters is higher education than Bachelors; use your best guess in ordering those levels for which the order is not obvious). Create a bar chart visualising the relative frequencies of values.**

In [ ]:

data['education'] = data['education'].astype('category')
#data['education'].cat.reorder_categories(['Doctorate','Masters',
 #                                         'Bachelors','Some-college',
  #                                        'Prof-school','Assoc-voc',
   #                                       'Assoc-acdm','HS-Grad','12th',
    #                                      '11th','10th','9th',
     #                                     '7th-8th','5th-6th','1st-4th','Preschool'
      #                                   ])


data['education'].value_counts(normalize=True).reindex(['Doctorate','Masters',
                                          'Bachelors','Some-college',
                                          'Prof-school','Assoc-voc',
                                          'Assoc-acdm','HS-Grad','12th',
                                          '11th','10th','9th',
                                          '7th-8th','5th-6th','1st-4th','Preschool'
                                         ]).plot(kind="bar")




**4.2.  Create and print a Pandas data frame where each row stands for an occupation, each column stands for a level of education, and the cells in the table contain the average salary of people with the corresponding occupation and education level.** <br> 
Hint 1: You can use crosstab, see also aggfunc from the pandas documentation.

In [4]:
# TODO
import numpy as np
import pandas as pd

crosstab_df = pd.crosstab(data.occupation, data.education, values = data.salaries, aggfunc=np.mean).round(2)
crosstab_df

education               10th      11th      12th   1st-4th   5th-6th  \
occupation                                                             
 ?                  31292.25  29582.05  31123.75  30193.58  31864.57   
 Adm-clerical       29957.97  29976.85  29722.61       NaN  26075.00   
 Armed-Forces            NaN       NaN  48635.00       NaN       NaN   
 Craft-repair       35271.21  36050.30  35800.19  32884.09  33559.63   
 Exec-managerial    38655.33  36805.44  36365.31  50014.50  71742.00   
 Farming-fishing    32535.09  34774.05  33999.88  31450.00  30814.36   
 Handlers-cleaners  32047.00  30906.23  32224.45  31633.06  32748.48   
 Machine-op-inspct  32183.33  31165.66  28159.49  32227.96  33421.93   
 Other-service      30484.64  31202.76  31881.76  29770.65  31433.08   
 Priv-house-serv    33172.00  27458.43  34069.25  25066.36  26133.07   
 Prof-specialty     45071.67  30745.05  37524.00  23128.50  39091.00   
 Protective-serv    26930.17  38363.14  30582.83  41685.00  19378.00   
 Sales              33370.27  31411.95  35273.38  31324.25  39858.67   
 Tech-support       44464.67  32796.67  43313.00       NaN  39695.00   
 Transport-moving   38104.95  32254.36  39585.77  33409.88  31557.00   

education            7th-8th       9th   Assoc-acdm   Assoc-voc   Bachelors  \
occupation                                                                    
 ?                  32282.27  31683.63     34653.87    36020.44    40028.50   
 Adm-clerical       35226.91  32473.00     34836.39    33989.10    38987.47   
 Armed-Forces            NaN       NaN          NaN         NaN    15014.00   
 Craft-repair       33048.06  35322.85     41023.08    42391.51    45141.61   
 Exec-managerial    42949.74  38819.62     46605.03    45784.69    50908.82   
 Farming-fishing    33664.48  31886.18     32939.19    38431.48    39866.14   
 Handlers-cleaners  30780.35  34131.20     33321.33    34322.86    37371.94   
 Machine-op-inspct  33776.59  32178.88     43272.30    39073.59    38791.84   
 Other-service      31001.63  31044.70     33145.14    33618.78    35861.41   
 Priv-house-serv    28921.25  33230.20     16753.50    34781.25    37750.43   
 Prof-specialty     28300.22  35095.00     40030.98    42809.15    44600.97   
 Protective-serv    30010.33  33974.00     42760.50    41284.54    50673.92   
 Sales              37247.90  33974.47     40032.57    38573.94    47796.29   
 Tech-support       32947.20  32152.50     40934.41    42085.61    43159.54   
 Transport-moving   35341.45  33625.57     32465.22    41845.22    40176.13   

education            Doctorate   HS-grad   Masters   Preschool   Prof-school  \
occupation                                                                     
 ?                    57649.47  33286.82  43630.42    24989.20      46083.56   
 Adm-clerical         43041.60  34729.64  42550.85    31927.50      44568.11   
 Armed-Forces              NaN  35544.00  67993.00         NaN           NaN   
 Craft-repair         54458.50  38615.12  46957.59    33937.50      52143.71   
 Exec-managerial      63967.98  42182.16  57455.88         NaN      58638.12   
 Farming-fishing      70039.00  35043.57  40793.10    32559.33      43915.50   
 Handlers-cleaners         NaN  32552.26  29983.00    30799.00           NaN   
 Machine-op-inspct    70502.00  35660.08  31144.50    32628.82      38026.00   
 Other-service        71626.00  32137.30  32017.68    32498.47      37343.75   
 Priv-house-serv           NaN  31240.14  28568.00    26577.00           NaN   
 Prof-specialty       56764.62  40999.92  48030.88    34861.00      58333.60   
 Protective-serv           NaN  39481.47  48472.13         NaN      64038.00   
 Sales                53759.50  37536.60  52718.71         NaN      52483.89   
 Tech-support         58249.00  40867.11  48925.41         NaN      42851.86   
 Transport-moving     18726.00  38235.36  43370.40         NaN      67293.33   

education            Some-college  
occupation                        

**4.3. Make a separate plot for each row (occupation) of the table created in subtask b. Choose the type of the plot to convey the information as well as you can.** <br>

In [ ]:
if crosstab_df is not crosstab_df.empty:
    for index, row in crosstab_df.iterrows():
        row.plot(kind='bar') # TODO 
        print(index)
        plt.show()

**4.4. List 3 interesting facts that you can read out of these plots from subtask c.** <br>

**<font color='red'>Fact 1:</font>** For occupation armed-forces it shows that for people who have education 12th grade have higher salaries than people with bachelors, some-college. and HS-grade background.

**<font color='red'>Fact 2:</font>**  For occupation exec-managerial it shows that for people with education 5th-6th grade have the highest salaries compare to others.

**<font color='red'>Fact 3:</font>**  For occupation farming-fishing the highest salaries is owned by people with 'Doctorate' education

## How long did it take you to solve the homework?

* Please answer as precisely as you can. It does not affect your points or grade in any way. It is okey, if it took 0.5 hours or 24 hours. The collected information will be used to improve future homeworks.

<font color='red'> **Answer:**</font>

**<font color='red'>(please change X in the next cell into your estimate)</font>**

6-7 hours

## How much the homework is difficult?
you can put only number between $0:10$ ($0:$ easy, $10:$ difficult)

<font color='red'> **Answer: 8**</font>


